### PR TITLE
Protocol changes for 1.21.20

### PR DESCRIPTION
**Pull request by dries-c**

### CODE REVIEW

File: src/StopSoundPacket.php [hunk 1]
```diff
@@ -21,25 +21,29 @@ class StopSoundPacket extends DataPacket implements ClientboundPacket{
 
 	public string $soundName;
 	public bool $stopAll;
+	public bool $isStopLegacyMusic;
```
Comment (dktapps) [inline src/StopSoundPacket.php:24]: `stopLegacyMusic` would be more consistent

**Comment by PMK744** [comment]
There seems to be an additional field for the ItemUseInventoryTransaction type.

image

**Comment by AzOxStOz** [comment]
I use your version [dries-c](https://github.com/dries-c) but item don't show to player in inventory

**Comment by JblusItsMe** [comment]
You need to add a **dynamicContainerId** on **Inventorymanager** as well as **ItemStackResponseContainerInfo**, for now, **-1** and the only values I found that makes the inventory display, but I still have bugs when I select an item @AzOxStOz

**Comment by AzOxStOz** [comment]
yeah did that:

private function sendInventorySlotPackets(int $windowId, int $netSlot, ItemStackWrapper $itemStackWrapper) : void{
		/*
		 * TODO: HACK!
		 * As of 1.20.12, the client ignores change of itemstackID in some cases when the old item == the new item.
		 * Notably, this happens with armor, offhand and enchanting tables, but not with main inventory.
		 * While we could track the items previously sent to the client, that's a waste of memory and would
		 * cost performance. Instead, clear the slot(s) first, then send the new item(s).
		 * The network cost of doing this is fortunately minimal, as an air itemstack is only 1 byte.
		 */

		$inventory = $this->session->getPlayer()->getInventory();
		$dynamicContainerId = $this->getWindowId($inventory);

		if ($dynamicContainerId === null) {
			$dynamicContainerId = $this->addDynamic($inventory);
		}

		if($itemStackWrapper->getStackId() !== 0){
			$this->session->sendDataPacket(InventorySlotPacket::create(
				$windowId,
				$netSlot,
				new ItemStackWrapper(0, ItemStack::null()),
				$dynamicContainerId
			));
		}
		//now send the real contents
		$this->session->sendDataPacket(InventorySlotPacket::create(
			$windowId,
			$netSlot,
			$itemStackWrapper,
			$dynamicContainerId
		));
	}

**Comment by JblusItsMe** [comment]
https://github.com/user-attachments/assets/326e4903-ac1a-45f0-a286-cd0e7fd23a1a

I have already managed to correct the inventories, now, the last concern I have is the world, I must find

**Comment by Nya-Enzo** [comment]
in terms of the world, i'm pretty sure the solution is to update the blockstates nbt file that provides the runtime block map to the server

**Comment by AzOxStOz** [comment]
@JblusItsMe how do you managed to correct the inventories ? because i do something but when i selected an item, like i pick one the item it duplicate

**Comment by JblusItsMe** [comment]
@AzOxStOz 
```php
public function build() : ItemStackResponse{
		$responseInfosByContainer = [];
		foreach($this->changedSlots as $containerInterfaceId => $slotIds){
			if($containerInterfaceId === ContainerUIIds::CREATED_OUTPUT){
				continue;
			}
			foreach($slotIds as $slotId){
				$inventoryAndSlot = $this->getInventoryAndSlot($containerInterfaceId, $slotId);
				if($inventoryAndSlot === null){
					//a plugin may have closed the inventory during an event, or the slot may have been invalid
					continue;
				}
				[$inventory, $slot] = $inventoryAndSlot;

				$itemStackInfo = $this->inventoryManager->getItemStackInfo($inventory, $slot);
				if($itemStackInfo === null){
					throw new AssumptionFailedError("ItemStackInfo should never be null for an open inventory");
				}
				$item = $inventory->getItem($slot);

				$responseInfosByContainer[$containerInterfaceId][] = new ItemStackResponseSlotInfo(
					$slotId,
					$slotId,
					$item->getCount(),
					$itemStackInfo->getStackId(),
					$item->getCustomName(),
					$item instanceof Durable ? $item->getDamage() : 0,
				);
			}
		}

		$responseContainerInfos = [];
		foreach($responseInfosByContainer as $containerInterfaceId => $responseInfos){

// TODO: HERE
			$responseContainerInfos[] = new ItemStackResponseContainerInfo(new FullContainerName($containerInterfaceId), $responseInfos);
		}

		return new ItemStackResponse(ItemStackResponse::RESULT_OK, $this->requestId, $responseContainerInfos);
	}
```

In the file **ItemStackResponseBuilder.php**, **line 103**, when calling **ItemStackResponseContainerInfo**, it asks for a **FullContainerName**, on the version of pocketmine, it is written **$containerInterfaceId**, you have to replace it with new **FullContainerName($containerInterfaceId)**

**Comment by AzOxStOz** [comment]
thank you i try

**Comment by AzOxStOz** [comment]
that's work fine thanks, i haved created a parameter before and somehow it didnt work

**Comment by AzOxStOz** [comment]
@JblusItsMe when i leave and reconnect the server my item aren't show but they are in my inventory, you too ?

**Comment by JblusItsMe** [comment]
https://youtu.be/y_UoRUDAZbU?si=b_n2HfXw1JWRGadH

@AzOxStOz here, everything is patch, I build and send it here

**Comment by AzOxStOz** [comment]
aw thanks you bro, i've got some big probleme with stone_bricks afterwards with the bedrock upgarde schema i don't know if you too @JblusItsMe

**Comment by JblusItsMe** [comment]
@AzOxStOz I have a new concern, the world this charge well, I see my blocks, but now I have update_block that replaces the dirt, I will correct it
image

**Comment by AzOxStOz** [comment]
@JblusItsMe yeah that's in the bedrock block upgrade schema as i mentioned before, their is a problem with "dirt_type", but if you replace it world edit it works fine, that's strange

**Comment by ShockedPlot7560** [comment]
Can you moove into discord or smth and resume here when you're done ?

**Comment by fBNTFeujjj** [comment]
@AzOxStOz Tu parles avec @JblusItsMe en anglais, mais vous êtes français mdr

**Comment by AzOxStOz** [comment]
ptet parcqu'on est dans une pr ?

**Comment by IvanCraft623** [comment]
@fBNTFeujjj @AzOxStOz @JblusItsMe Move your conversation elsewhere and it's make difficult for those who are interested in reviewing the PR

**Comment by IvanCraft623** [comment]
Please don't force-push

**Review by CJMustard1452 — APPROVED** [review]
this looks very good dries-c

File: src/types/inventory/stackrequest/CraftRecipeAutoStackRequestAction.php [hunk 2]
```diff
@@ -34,13 +34,16 @@ final class CraftRecipeAutoStackRequestAction extends ItemStackRequestAction{
...
 		private int $recipeId,
+		private int $repetitions2,
 		private int $repetitions,
 		private array $ingredients
 	){}
 
 	public function getRecipeId() : int{ return $this->recipeId; }
 
-	public function getRepetitions() : int{ return $this->repetitions; }
+	public function getRepetitions() : int { return $this->repetitions; }
```
Comment (nicholass003) [inline src/types/inventory/stackrequest/CraftRecipeAutoStackRequestAction.php:44]: remove space for code consistency
```suggestion
	public function getRepetitions() : int{ return $this->repetitions; }
```

File: src/types/inventory/stackrequest/CraftRecipeStackRequestAction.php [hunk 3]
```diff
@@ -26,17 +26,22 @@ final class CraftRecipeStackRequestAction extends ItemStackRequestAction{
...
 
 	final public function __construct(
-		private int $recipeId
+		private int $recipeId,
+		private int $repetitions
 	){}
 
 	public function getRecipeId() : int{ return $this->recipeId; }
 
+	public function getRepetitions() : int { return $this->repetitions; }
```
Comment (nicholass003) [inline src/types/inventory/stackrequest/CraftRecipeStackRequestAction.php:35]: same here
```suggestion
	public function getRepetitions() : int{ return $this->repetitions; }
```

File: src/types/inventory/stackrequest/CreativeCreateStackRequestAction.php [hunk 4]
```diff
@@ -26,17 +26,22 @@ final class CreativeCreateStackRequestAction extends ItemStackRequestAction{
...
 
 	public function __construct(
-		private int $creativeItemId
+		private int $creativeItemId,
+		private int $repetitions
 	){}
 
 	public function getCreativeItemId() : int{ return $this->creativeItemId; }
 
+	public function getRepetitions() : int { return $this->repetitions; }
```
Comment (nicholass003) [inline src/types/inventory/stackrequest/CreativeCreateStackRequestAction.php:35]: same here
```suggestion
	public function getRepetitions() : int{ return $this->repetitions; }
```

File: src/types/inventory/stackrequest/GrindstoneStackRequestAction.php [hunk 5]
```diff
@@ -27,23 +27,28 @@ final class GrindstoneStackRequestAction extends ItemStackRequestAction{
...
+		private int $repairCost,
+		private int $repetitions
 	){}
 
 	public function getRecipeId() : int{ return $this->recipeId; }
 
 	/** WARNING: This may be negative */
 	public function getRepairCost() : int{ return $this->repairCost; }
 
+	public function getRepetitions() : int { return $this->repetitions; }
```
Comment (nicholass003) [inline src/types/inventory/stackrequest/GrindstoneStackRequestAction.php:39]: same here
```suggestion
	public function getRepetitions() : int{ return $this->repetitions; }
```

**Review by nicholass003 — CHANGES_REQUESTED** [review]

File: src/JigsawStructureDataPacket.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace pocketmine\network\mcpe\protocol;
+
+use pocketmine\network\mcpe\protocol\serializer\PacketSerializer;
+use pocketmine\network\mcpe\protocol\types\CacheableNbt;
+
+class JigsawStructureDataPacket extends DataPacket implements ClientboundPacket{
+	public const NETWORK_ID = ProtocolInfo::JIGSAW_STRUCTURE_DATA_PACKET;
+
+	/** @phpstan-var CacheableNbt<\pocketmine\nbt\tag\CompoundTag> */
```
Comment (ShockedPlot7560) [inline src/JigsawStructureDataPacket.php:23]: Shouldn't generics be imported?

File: src/JigsawStructureDataPacket.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	private CacheableNbt $nbt;
+
+	/**
+	 * @generate-create-func
+	 * @phpstan-param CacheableNbt<\pocketmine\nbt\tag\CompoundTag> $nbt
```
Comment (ShockedPlot7560) [inline src/JigsawStructureDataPacket.php:28]: same

File: src/JigsawStructureDataPacket.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+	 */
+	public static function create(CacheableNbt $nbt) : self{
+		$result = new self;
+		$result->nbt = $nbt;
+		return $result;
+	}
+
+	/** @phpstan-return CacheableNbt<\pocketmine\nbt\tag\CompoundTag> */
```
Comment (ShockedPlot7560) [inline src/JigsawStructureDataPacket.php:36]: same

Comment (dries-c) [reply src/JigsawStructureDataPacket.php:36]: It's like this in all packets except for the StartGamePacket, so the majority of packets use the long form

Comment (ShockedPlot7560) [reply src/JigsawStructureDataPacket.php:36]: I didn't know about it, all right.

**Review by ShockedPlot7560 — APPROVED** [review]
Looks ok to me

**Review by dktapps — APPROVED** [review]

**Review by IvanCraft623 — CHANGES_REQUESTED** [review]
`CraftRecipeOptionalStackRequestAction` needs a `repetitions` field
`CorrectPlayerMovePredictionPacket` needs `vehicleAngularVelocity` field when type is PREDICTION_TYPE_VEHICLE

**Pull request merged**